### DISCUSSION TITLE
Syntax conflict between knit_expand and Quarto extensions?

### DISCUSSION THREAD

**Discussion by daranzolin** [quarto-dev/quarto-cli#4286] [Q&A] [answered]
I'm having difficulty using Quarto extensions within a child document, and I'm wondering if there's a syntax conflict with the curly braces. For example, knit_expand uses "{{i}}" to reference a variable, and Quarto extensions start and end with '{...}' I'll try and spin up a reprex at some point, but the error I'm getting is:

```
Error: Failed to parse the inline R code: < ...... (Reason: <text>:1:1: unexpected '<'
1: <
    ^)
```

**Comment by cscheid** [comment]
We'll unfortunately need a full example before being able to help.

**Comment by daranzolin** [comment, marked as answer]
[Ok, I created a gist here.](https://gist.github.com/daranzolin/bcf458ccfe0caa65efd4b84854508389)

The error:
```
Quitting from lines 8-13 (template.qmd) 
Error: Failed to parse the inline R code: < downloadthis ", "'", path, "'", ' dname=Mtcars label="Download the mtcars data" icon=database-fill-down type=info class=data-button id=mtcars > (Reason: <text>:1:1: unexpected '<'
1: <
    ^)

Execution halted
```

Edit: 'mtcars.csv' must be in the same project directory, sorry.

Edit2: I've tried using a couple other Quarto extensions within child.qmd and none of them have worked.

**Reply by baptiste** [reply]
I don't quite understand what you want to replace in the template; if I use a valid R string like this in child.qmd,

````
```{r}
cmd <- "The value of `pi` is < {{pi}} >."
```

```{r}
#| results: asis
cat(cmd)
```

````

then everything works as intended. On the other hand, with your example,

```
cmd <- paste0("{{< downloadthis ", "'", path, "'", ' dname=Mtcars label="Download the mtcars data" icon=database-fill-down type=info class=data-button id=mtcars >}}')
```

R cannot execute what's inside the pair of `{{` `}}`, because it's not a valid R expression.

**Reply by daranzolin** [reply]
@baptiste [many Quarto extensions, including downloadthis, use this double curly brace convention.](https://github.com/shafayetShafee/downloadthis) I want Quarto/knitr to evaluate and execute the extension syntax.

**Reply by baptiste** [reply]
I still don't get what you want to expand in the knit_expand template (maybe nothing in your example?), but as far as the curly-brace convention is concerned, you can simply use different delimiters, e.g

```
src = knitr::knit_expand("child.qmd", delim = c("<<", ">>"))
```

_Edit:_ I've [forked your gist](https://gist.github.com/baptiste/ddd8fe83f9413001bb96e56178181dff) with an example using different delimiters, and expanding a variable in the template

**Reply by daranzolin** [reply]
@baptiste that works, thank you!!!! Didn't know about the delim argument. I *think* it also confirms my initial hunch that `knit_expand()` and the quarto extension syntax were in conflict.

**Reply by mcanouil** [reply]
For the sake of clarity, there is no `{{ ... }}` convention in Quarto.
It's shortcode syntax which is `{{< ... >}}`.

**Reply by cderv** [reply]
To complete what @baptiste said, I don't think you need `knit_expand()` here, as there is nothing to expand. 

I don't think we should consider conflict here. It is just that if you need to use `knit_expand()` with some content that are using the default delimiters as not-delimiter to R code to evaluate, then the `delim` argument should be used. **knitr** feature found int the function help page. The issue would be the same with **glue** package to build some Markdown string for Quarto using brackets. 

Regarding your gist, it could be done with `knit_child()` and without a `child.qmd`, but I assume this was a minimal reprex of a larger usage. 
here is an example where `knit_expand` is used to replace `path` in the string to build a shortcode. 

``` r
path <- fs::path("mtcars", ext = "csv")
src = knitr::knit_expand(
  text = "{{< downloadthis '<< path >>' dname=Mtcars label=\"Download the mtcars data\" icon=database-fill-down type=info class=data-button id=mtcars >}}",
  delim = c("<<", ">>")
)
src
#> [1] "{{< downloadthis 'mtcars.csv' dname=Mtcars label=\"Download the mtcars data\" icon=database-fill-down type=info class=data-button id=mtcars >}}"
```

<sup>Created on 2023-02-09 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

And below is an example to create a function that does output the built string to be include asis by **knitr**

````markdown
```{r, include = FALSE}
download_this <- function(path) {
  src <- knitr::knit_expand(
    text = "{{< downloadthis '<< path >>' dname=Mtcars label=\"Download the mtcars data\" icon=database-fill-down type=info class=data-button id=mtcars >}}",
    delim = c("<<", ">>")
  )
  knitr::asis_output(src)
}
```

```{r}
download_this("mtcars.csv")
```

````


hope it helps understand

**Reply by daranzolin** [reply]
@cderv thanks, that's also really helpful!
